{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "prepareLongestCDStranscript (generic function with 1 method)"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import the GFF3 module.\n",
    "using GFF3\n",
    "using DataStructures\n",
    "\n",
    "function prepareLongestCDStranscript(inputFile,outputFile)\n",
    "    #####################################################\n",
    "    # 1.get exon CDS and gene info\n",
    "    #####################################################\n",
    "    # feature type info\n",
    "    geneinfoDict = Dict{String,String}()\n",
    "    mRNAinfoDict = Dict{String,String}()\n",
    "    exonDict = Dict{String,Int64}()\n",
    "    CDSDict = Dict{String,Int64}()\n",
    "\n",
    "    # Open a GFF3 file.\n",
    "    reader = open(GFF3.Reader, inputFile)\n",
    "\n",
    "    # Iterate over records.\n",
    "    for record in reader\n",
    "        # save gene_id and gene name\n",
    "        if GFF3.featuretype(record) == \"gene\"\n",
    "            gene_id = GFF3.attributes(record,\"gene_id\")[1]\n",
    "            try\n",
    "                geneinfoDict[gene_id] = GFF3.attributes(record,\"Name\")[1]\n",
    "            catch\n",
    "                geneinfoDict[gene_id] = \"non-name\"\n",
    "            end\n",
    "        end\n",
    "        # save transcript_id and gene_id\n",
    "        if GFF3.featuretype(record) == \"mRNA\"\n",
    "            transcript_id = GFF3.attributes(record,\"transcript_id\")[1]\n",
    "            gene_id = split(GFF3.attributes(record,\"Parent\")[1],\":\")[2]\n",
    "            mRNAinfoDict[transcript_id] = gene_id\n",
    "        end\n",
    "        # sum up exon length\n",
    "        if GFF3.featuretype(record) == \"exon\"\n",
    "            transid = split(GFF3.attributes(record,\"Parent\")[1],\":\")[2]\n",
    "            exonLength = abs(GFF3.seqend(record) - GFF3.seqstart(record)) + 1\n",
    "            if !haskey(exonDict,transid)\n",
    "                exonDict[transid] = exonLength\n",
    "            else\n",
    "                exonDict[transid] += exonLength\n",
    "            end\n",
    "        end\n",
    "        # sum up CDS length\n",
    "        if GFF3.featuretype(record) == \"CDS\"\n",
    "            transid = split(GFF3.attributes(record,\"Parent\")[1],\":\")[2]\n",
    "            cdsLength = abs(GFF3.seqend(record) - GFF3.seqstart(record)) + 1\n",
    "            # println(exonLength)\n",
    "            if !haskey(CDSDict,transid)\n",
    "                CDSDict[transid] = cdsLength\n",
    "            else\n",
    "                CDSDict[transid] += cdsLength\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "\n",
    "    # Finally, close the reader\n",
    "    close(reader)\n",
    "\n",
    "    #####################################################\n",
    "    # 2.combine gene and and trascript info\n",
    "    #####################################################\n",
    "    # merge info\n",
    "    merExonDict = DefaultDict(Dict)\n",
    "    merCDSDict = DefaultDict(Dict)\n",
    "\n",
    "    # combine gene info with exon length\n",
    "    for (tranid,exonlength) in exonDict\n",
    "        if haskey(mRNAinfoDict,tranid)\n",
    "            key1 = geneinfoDict[mRNAinfoDict[tranid]]*\"|\"*mRNAinfoDict[tranid]\n",
    "            merExonDict[key1][tranid] = exonlength\n",
    "        end\n",
    "    end\n",
    "    # combine gene info with cds length\n",
    "    for (tranid,cdslength) in CDSDict\n",
    "        if haskey(mRNAinfoDict,tranid)\n",
    "            key1 = geneinfoDict[mRNAinfoDict[tranid]]*\"|\"*mRNAinfoDict[tranid]\n",
    "            merCDSDict[key1][tranid] = cdslength\n",
    "        end\n",
    "    end\n",
    "\n",
    "    #####################################################\n",
    "    # 3.filter longsest cds-exon transcript\n",
    "    #####################################################\n",
    "    # save final longest CDS transcript info\n",
    "    finalResDict = Dict{String,Vector{Int64}}()\n",
    "\n",
    "    # filter longest CDS-exon transcript\n",
    "    for (ginfo,cdsinfo) in merCDSDict\n",
    "        maxcds = maximum(values(cdsinfo)) # max cds length\n",
    "        cdslongestDict = Dict{String,Int64}() # save in dict\n",
    "        for (key,val) in cdsinfo\n",
    "            if val == maxcds\n",
    "                cdslongestDict[key] = val\n",
    "            end\n",
    "        end\n",
    "        # whether have one cds longest transcript\n",
    "        if length(keys(cdslongestDict)) == 1\n",
    "            longestTranid = [i for i in keys(cdslongestDict)][1]\n",
    "            # save\n",
    "            finalResDict[\"$ginfo|$longestTranid\"] = [merExonDict[ginfo][longestTranid],maxcds]\n",
    "        # which have equal CDS length\n",
    "        elseif length(keys(cdslongestDict)) > 1\n",
    "            tmpexondict = merExonDict[ginfo]\n",
    "            exonlongestDict = Dict{String,Int64}() # save in dict\n",
    "            for key in keys(cdslongestDict)\n",
    "                if haskey(tmpexondict,key)\n",
    "                    exonlongestDict[key] = tmpexondict[key]\n",
    "                end\n",
    "            end\n",
    "            maxexon = maximum(values(exonlongestDict)) # max exon length\n",
    "            for (key,val) in exonlongestDict\n",
    "                if val == maxexon\n",
    "                    # save\n",
    "                    finalResDict[\"$ginfo|$key\"] = [merExonDict[ginfo][key],maxcds]\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "\n",
    "    #####################################################\n",
    "    # 4.save gene other info like chromsome strand and CDS regions\n",
    "    #####################################################\n",
    "    longestidTrnscript = Dict{String,String}()\n",
    "    for key in keys(finalResDict)\n",
    "        tid = split(key,\"|\")[3]\n",
    "        longestidTrnscript[tid] = key\n",
    "    end\n",
    "    # save gene other info\n",
    "    outcdsDict = Dict{String,Vector{String}}()\n",
    "    outexonDict = Dict{String,Vector{String}}()\n",
    "    UTR5Dict = Dict{String,Int64}()\n",
    "\n",
    "    # Open a GFF3 file.\n",
    "    reader = open(GFF3.Reader, inputFile)\n",
    "\n",
    "    # Iterate over records.\n",
    "    for record in reader\n",
    "        # filter longest transcript gene info\n",
    "        if GFF3.featuretype(record) == \"CDS\"\n",
    "            transid = split(GFF3.attributes(record,\"Parent\")[1],\":\")[2]\n",
    "            if haskey(longestidTrnscript,transid)\n",
    "                # tags\n",
    "                id = longestidTrnscript[transid]\n",
    "                chrom = GFF3.seqid(record)\n",
    "                strand = GFF3.strand(record)\n",
    "                seqstart = GFF3.seqstart(record)\n",
    "                seqend = GFF3.seqend(record)\n",
    "                key = \"$id|$chrom|$strand\"\n",
    "                region = \"$seqstart:$seqend\"\n",
    "                # add multiple cds region\n",
    "                if !haskey(outcdsDict,key)\n",
    "                    outcdsDict[key] = [region]\n",
    "                else\n",
    "                    push!(outcdsDict[key],region)\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "        # filter longest transcript gene info\n",
    "        if GFF3.featuretype(record) == \"exon\"\n",
    "            transid = split(GFF3.attributes(record,\"Parent\")[1],\":\")[2]\n",
    "            if haskey(longestidTrnscript,transid)\n",
    "                # tags\n",
    "                id = longestidTrnscript[transid]\n",
    "                chrom = GFF3.seqid(record)\n",
    "                strand = GFF3.strand(record)\n",
    "                seqstart = GFF3.seqstart(record)\n",
    "                seqend = GFF3.seqend(record)\n",
    "                key = \"$id|$chrom|$strand\"\n",
    "                region = \"$seqstart:$seqend\"\n",
    "                # add multiple exon region\n",
    "                if !haskey(outexonDict,key)\n",
    "                    outexonDict[key] = [region]\n",
    "                else\n",
    "                    push!(outexonDict[key],region)\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "        # get 5utr length\n",
    "        if GFF3.featuretype(record) == \"five_prime_UTR\"\n",
    "            transid = split(GFF3.attributes(record,\"Parent\")[1],\":\")[2]\n",
    "            if haskey(longestidTrnscript,transid)\n",
    "                utr5Length = abs(GFF3.seqend(record) - GFF3.seqstart(record)) + 1\n",
    "                if !haskey(UTR5Dict,transid)\n",
    "                    UTR5Dict[transid] = utr5Length\n",
    "                else\n",
    "                    UTR5Dict[transid] += utr5Length\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "\n",
    "    # Finally, close the reader\n",
    "    close(reader)\n",
    "\n",
    "    #####################################################\n",
    "    # 5.finally output tab-delimit file\n",
    "    # fullid gene_id transcript_id chromosome strand CDS_regions exon_regions 5UTR_length CDS_length 3UTR_length\n",
    "    #####################################################\n",
    "    # OR4F5|ENSG00000186092|ENST00000641515|1|+\n",
    "    outputfile = open(outputFile,\"w\")\n",
    "\n",
    "    # output file\n",
    "    for(key,val) in outcdsDict\n",
    "        gene_name,gene_id,transid,chrom,strand = split(key,\"|\")\n",
    "        cds = join(val,\",\")\n",
    "        exon = join(outexonDict[key],\",\")\n",
    "        if haskey(UTR5Dict,transid) # utr5 length\n",
    "            utr5length = UTR5Dict[transid] \n",
    "        else\n",
    "            utr5length = 0\n",
    "        end\n",
    "        cdslength = CDSDict[transid] # cds length\n",
    "        utr3length = exonDict[transid] - (utr5length + cdslength) # utr3 length\n",
    "        write(outputfile,join([key,gene_name,gene_id,transid,chrom,strand,cds,exon,utr5length,cdslength,utr3length],\"\\t\")*\"\\n\")\n",
    "    end\n",
    "    # close file\n",
    "    close(outputfile)\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# prepareLongestCDStranscript(\"Homo_sapiens.GRCh38.106.gff3\",\"longestCDSinfo.txt\")\n",
    "prepareLongestCDStranscript(\"Homo_sapiens.GRCh38.106.gff3\",\"longestCDSinfo.txt\")"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "e26550a99b8cdaeca938eba5dfe93bbe201b1984deb8af224ca0d45a986faa87"
  },
  "kernelspec": {
   "display_name": "Julia 1.7.2",
   "language": "julia",
   "name": "julia-1.7"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.7.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
